master/reference/advanced/etcd-rbac/certificate-generation.md

```diff
@@ -0,0 +1,92 @@
+to make certificate generation easy.  If you use hack/tls-setup too, ensure
+you have followed the README and are able to run `make` successfully.
+
+It is important to ensure that the Common Name (CN), of each component that
+will be accessing etcd, match the username in etcd that has the appropriate
+etcd roles for accessing its needed keys or paths.
+
+> **Note:** Re-running `make` using hack/tls-setup will regenerate the CA and
+all certificates.
+
+## Generating Certificate Authority
+
+The default CA configuration included with hack/tls-setup works well with no
+additional configuration.  The file `certs/ca.pem` generated will need to be
+provided to all components (etcd, Kubernetes apiserver, and all calico
+components).
+
+## Generating etcd certificates
+
+The hack/tls-setup is meant to generate etcd certificates so the file
+`config/req-csr.json` needs to be updated by adding the IP addresses of the
+servers that will be running the etcd members added to the `"hosts"`
+section.  After that, running `make` will generate three certs that can be
+used for three etcd member servers (though just using one works, when testing).
+These cert and key files are `certs/etcd[123].pem` and
+`certs/etcd[123]-key.pem` and a matching pair will need to be provided to
+each etcd member.
+
+If using etcd proxies, the cert/key pair generated by the tool
+(the files `certs/proxy1.pem` and `certs/proxy1-key.pem`) can be used with all
+proxies or you could create individual cert/key pairs for each proxy too.
+
+## Generating per-user/per-component's etcd certificates
+
+The certificates for Calico, Kubernetes, or any other component can be
+generated with configuration files similar to the one provided below.
+The `"CN"` name (`etcd_username` in the example below) must be udpated with
+the username that will be created in etcd and given the proper roles which will
+allow access to the needed paths/prefixed keys that the components need.
+
+```
+{ 
+  "CN": "etcd_username",
+  "hosts": [
+    "localhost"
+  ],
+  "key": {
+    "algo": "ecdsa",           
+    "size": 384                
+  },  
+  "names": [
+    { 
+      "O": "autogenerated",    
+      "OU": "etcd cluster",
+      "L": "the internet"
+    } 
+  ]   
+}
+```
+
+The additional configuration files you create should be added to the `config`
+directory located in your hack/tls-setup folder.  To build certificates for
+each new configuration add lines similar to those below to
+the `req:` target in the Makefile.  For each configuration added, make sure the
+configuration file name (`config/req-component.json`) and cert/key file prefix
+(`certs/component`) are updated appropriately.
+
+```
+    $(CFSSL) gencert \
+      -ca certs/ca.pem \
+      -ca-key certs/ca-key.pem \
+      -config config/ca-config.json \
+      config/req-component.json | $(JSON) -bare certs/component
+```
+
+Once you have done all the configuration needed above then a simple `make` will
```

Comment:
I think this last bit about running the script should be a separate section, e.g. "Generating the certificates and keys"